phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -0,0 +1,315 @@
+**                          |/____/                                         **
+\*                                                                          */
+
+package org.scalajs.jsenv.phantomjs
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Promise, Future}
+import scala.util.control.NonFatal
+
+import java.io._
+import java.net._
+import java.nio.file.{Files, StandardCopyOption}
+
+import org.scalajs.jsenv._
+
+import org.scalajs.io._
+import org.scalajs.io.URIUtils.fixFileURI
+import org.scalajs.io.JSUtils.escapeJS
+
+private final class ComRun(jettyClassLoader: ClassLoader,
+    runConfig: RunConfig, onMessage: String => Unit,
+    startRun: VirtualBinaryFile => JSRun)
+    extends JSComRun {
+
+  import ComRun._
+
+  import runConfig.logger
+
+  private var state: State = WaitingForServerToStart(Nil)
+
+  private[this] val fragmentsBuf = new StringBuilder
+
+  private val promise = Promise[Unit]()
+
+  def future: Future[Unit] = promise.future
+
+  private def loadMgr(): WebsocketManager = {
+    val loader =
+      if (jettyClassLoader != null) jettyClassLoader
+      else getClass().getClassLoader()
+
+    val clazz = loader.loadClass(
+        "org.scalajs.jsenv.phantomjs.JettyWebsocketManager")
+
+    val ctors = clazz.getConstructors()
+    assert(ctors.length == 1, "JettyWebsocketManager may only have one ctor")
+
+    val listener = new WebsocketListener {
+      def onRunning(): Unit = onServerRunning()
+      def onOpen(): Unit = onConnectionOpened()
+      def onClose(): Unit = ()
+      def onMessage(msg: String): Unit = receiveFrag(msg)
+      def log(msg: String): Unit = logger.debug(s"PhantomJS WS Jetty: $msg")
+    }
+
+    val mgr = ctors.head.newInstance(listener)
+
+    mgr.asInstanceOf[WebsocketManager]
+  }
+
+  private val mgr: WebsocketManager = loadMgr()
+
+  // Constructor
+  mgr.start()
+
+  private def onServerRunning(): Unit = synchronized {
+    import ExecutionContext.Implicits.global
+
+    state match {
+      case WaitingForServerToStart(sendQueue) =>
+        state = AwaitingConnection(sendQueue)
+        val comSetup = makeComSetupFile()
+        val underlyingRun = startRun(comSetup)
+        underlyingRun.future.onComplete { result =>
+          promise.tryComplete(result)
+        }
+
+      case Closing =>
+        promise.trySuccess(())
+
+      case AwaitingConnection(_) | Connected =>
+        throw new IllegalStateException(
+            s"Illegal state in onServerRunning: $state")
+    }
+  }
+
+  private def onConnectionOpened(): Unit = synchronized {
+    state match {
+      case AwaitingConnection(sendQueue) =>
+        sendQueue.reverse.foreach(sendNow)
+        state = Connected
+
+      case Closing =>
+        mgr.stop()
+
+      case WaitingForServerToStart(_) | Connected =>
+        throw new IllegalStateException(
+            s"Illegal state in onConnectionOpened: $state")
+    }
+  }
+
+  private def makeComSetupFile(): VirtualBinaryFile = {
```

Comment:
Maybe pass `serverPort` as a parameter so you can move this out of the class into the companion?